### ISSUE TITLE
help request: Cannot find upstream nodes in Kubernetes discovery

### ISSUE THREAD

**Issue by liuzhiyong1982** [apache/apisix#10251]
### Description

I deployed the APISIX from official Helm apisix chart v2.2.0.  Below image is the overlay values for enabling Kubernetes service discovery.
![image](https://github.com/apache/apisix/assets/73316431/4dfccab3-8a8e-441a-8614-5da1405f4cee)

what I have done:
- created SA, enabled RBAC;
- provided two env (KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT) for Nginx; I have tried FQDN kubernetes.default.svc.cluster.local and service IP in the KUBERNETES_SERVICE_HOST env, but no luck;
- used correct Service Name format in Apisix Dashboard;
![image](https://github.com/apache/apisix/assets/73316431/9d2c3bc6-b71c-4755-9791-abc98bfa1e9c)

however, I still got 503 when I attempt to open the site and was told 'failed to set upstream: no valid upstream node: nil'
![image](https://github.com/apache/apisix/assets/73316431/23c2adc9-7257-45a3-b638-644a0e5eb9c0)

Did I missed something?

### Environment

- APISIX version (run `apisix version`): apache/apisix:3.5.0-debian
- Operating system (run `uname -a`): Azure Kubernetes Service
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`): nginx version: openresty/1.21.4.2

**Comment by shreemaan-abhishek** [comment]
Did you configure the service_name in the following format?

> function: nodes(service_name)
>
> description: nodes() function attempts to look up the ngx.shared.DICT for nodes corresponding to servicename, \ service_name should match pattern: [namespace]/[name]:[portName]_

ref: https://apisix.apache.org/docs/apisix/discovery/kubernetes/#single-cluster-mode-query-interface

**Comment by liuzhiyong1982** [comment]
@shreemaan-abhishek , yes, I am using the right serivce_name format in apisix upstream setting.

![image](https://github.com/apache/apisix/assets/73316431/e1d2890e-a222-4b28-bc37-264bd747a9eb)

![image](https://github.com/apache/apisix/assets/73316431/aff09e27-8167-4dfb-93b5-89b56731e9d1)

**Comment by monkeyDluffy6017** [comment]
@liuzhiyong1982 have you solved your problem?

**Comment by ray1888** [comment]
want help also, meeting the same issue

**Comment by mozhata** [comment]
met the same issue, the service not be set port name, then trying to use the port number, I get the same error log:

```
2024/05/16 09:14:23 [error] 386#386: *7955311 [lua] init.lua:548: handle_upstream(): failed to set upstream: no valid upstream node: nil, client: 10.77.101.0, server: _, request: "POST /xxx
```

**Comment by Thilip707** [comment]
pls give example format for this target port number instead port number

**Comment by luistilingue** [comment]
```
apiVersion: apisix.apache.org/v2
kind: ApisixUpstream
metadata:
  name: httpbin-upstream
  namespace: default
spec:
  loadbalancer:
    type: ewma
  discovery:
    type: kubernetes
    serviceName: default/httpbin:http
```

that way works for me.

**Comment by PasanBhanu** [comment]
I fixed the issue by following step.

Configurations are similar to @liuzhiyong1982 but I didn't add ngnix since my apisix is in the same cluster. 

for the service name use below format

namespace/service-name:port-name

Always use port name, the name you defined in the service.

Ex:

```
spec:
  ports:
    - name: http2
      protocol: TCP
      port: 80
      targetPort: 8081
```

Below is how my route config looks like 

![Image](https://github.com/user-attachments/assets/9ed61dab-d49b-4b51-971a-887e97510b82)

**Comment by Baoyuantop** [comment]
This issue is related to the older version of the dashboard. I will close this issue first. If there is still a problem, please open it again.
